checker/src/main/java/org/checkerframework/checker/mustcall/MustCallAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,363 @@
+   *     identical to anno
+   */
+  // Package private to permit usage from the visitor in the common assignment check.
+  /* package-private */ AnnotationMirror withoutClose(@Nullable AnnotationMirror anno) {
+    // shortcut for easy paths
+    if (anno == null || AnnotationUtils.areSame(anno, BOTTOM)) {
+      return BOTTOM;
+    } else if (!AnnotationUtils.areSameByName(
+        anno, "org.checkerframework.checker.mustcall.qual.MustCall")) {
+      return anno;
+    }
+    List<String> values =
+        AnnotationUtils.getElementValueArray(anno, mustCallValueElement, String.class);
+    if (!values.contains("close")) {
+      return anno;
+    }
+    return createMustCall(values.stream().filter(s -> !"close".equals(s)).toArray(String[]::new));
+  }
+
+  /**
+   * Returns true iff the given element represents a variable that was declared in a
+   * try-with-resources statement.
+   *
+   * @param elt an element; may be null, in which case this method always returns false
+   * @return true iff the given element represents a variable that was declared in a
+   *     try-with-resources statement
+   */
+  private boolean isDeclaredInTryWithResources(@Nullable Element elt) {
+    return elt != null && elt.getKind() == ElementKind.RESOURCE_VARIABLE;
+  }
+
+  /** Treat non-owning method parameters as @MustCallUnknown when the method is called. */
+  @Override
+  public void methodFromUsePreSubstitution(ExpressionTree tree, AnnotatedExecutableType type) {
+    ExecutableElement declaration;
+    if (tree instanceof MethodInvocationTree) {
+      declaration = TreeUtils.elementFromUse((MethodInvocationTree) tree);
+    } else if (tree instanceof MemberReferenceTree) {
+      declaration = (ExecutableElement) TreeUtils.elementFromTree(tree);
+    } else {
+      throw new BugInCF("unexpected type of method tree: " + tree.getKind());
+    }
+    changeNonOwningParameterTypesToTop(declaration, type);
+    super.methodFromUsePreSubstitution(tree, type);
+  }
+
+  @Override
+  protected void constructorFromUsePreSubstitution(
+      NewClassTree tree, AnnotatedExecutableType type) {
+    ExecutableElement declaration = TreeUtils.elementFromUse(tree);
+    changeNonOwningParameterTypesToTop(declaration, type);
+    super.constructorFromUsePreSubstitution(tree, type);
+  }
+
+  /**
+   * Changes the type of each parameter not annotated as @Owning to top. Also replaces the component
+   * type of the varargs array, if applicable.
+   *
+   * @param declaration a method or constructor declaration
+   * @param type the method or constructor's type
+   */
+  private void changeNonOwningParameterTypesToTop(
+      ExecutableElement declaration, AnnotatedExecutableType type) {
+    for (int i = 0; i < type.getParameterTypes().size(); i++) {
+      Element paramDecl = declaration.getParameters().get(i);
+      if (checker.hasOption(MustCallChecker.NO_LIGHTWEIGHT_OWNERSHIP)
```

Comment:
Doing so seems unnecessary, since `hasOption` already uses a cache of active options (i.e. each call after the first is just a table lookup). I would prefer avoiding writing and maintaining an unnecessary cache in this checker, for simplicity's sake. If you feel strongly, I can make the change, but I would do so reluctantly.

Reply:
@mernst, you requested a similar change in several other places in your latest round of comments, but never responded to this comment. Can you explain your reasoning as to why a separate boolean cache is needed for these particular options, even though `hasOption` already uses a cache? I don't understand how there could be any significant performance difference between the two, but your suggestion makes the code more complex and therefore harder to read.

Reply:
@kelloggm I didn't respond to the comment because it was hidden by GitHub on the webpage, because there were so many comments.  I did probably receive an email notification of it, but it was buried in an avalanche of not-very-interesting replies to comments.  Per https://homes.cs.washington.edu/~mernst/advice/github-pull-request.html#respond-to-reviews, could you please only respond to a comment if you disagree with it or need more information?  Then the email volume will be lower and it will be all signal, not mostly noise.  Thanks.

Reply:
I didn't realize `hasOption` is fast after the first call.  That's fine, then.